src/solvers/refinement/string_constraint_generator_float.cpp

```diff
@@ -0,0 +1,454 @@
+  mult.copy_to_operands(f);
+  mult.copy_to_operands(g);
+  mult.copy_to_operands(from_integer(rounding, unsignedbv_typet(32)));
+  return mult;
+}
+
+/// Convert integers to floating point to be used in operations with
+/// other values that are in floating point representation.
+/// \param i: an expression representing an integer
+/// \param spec: specification for floating point numbers
+/// \return An expression representing the value of the input
+///   integer as a float.
+exprt floatbv_of_int_expr(const exprt &i, const ieee_float_spect &spec)
+{
+  exprt rounding=from_integer(ieee_floatt::ROUND_TO_ZERO, unsignedbv_typet(32));
+  return floatbv_typecast_exprt(i, rounding, spec.to_type());
+}
+
+/// Estimate the decimal exponent that should be used to represent a given
+/// floating point value in decimal.
+/// We are looking for d such that n * 10^d = m * 2^e, so:
+/// d = log_10(m) + log_10(2) * e - log_10(n)
+/// m -- the magnitude -- should be between 1 and 2 so log_10(m)
+/// in [0,log_10(2)].
+/// n -- the magnitude in base 10 -- should be between 1 and 10 so
+/// log_10(n) in [0, 1].
+/// So the estimate is given by:
+/// d ~=~  floor(log_10(2) * e)
+/// \param f: a floating point expression
+/// \param spec: specification for floating point
+exprt estimate_decimal_exponent(const exprt &f, const ieee_float_spect &spec)
+{
+  exprt log_10_of_2=constant_float(0.30102999566398119521373889472449302, spec);
+  exprt bin_exp=get_exponent(f, spec);
+  exprt float_bin_exp=floatbv_of_int_expr(bin_exp, spec);
+  exprt dec_exp=floatbv_mult(float_bin_exp, log_10_of_2);
+  binary_relation_exprt negative_exp(dec_exp, ID_lt, constant_float(0.0, spec));
+  // Rounding to zero is not correct for negative numbers, so we substract 1.
+  minus_exprt dec_minus_one(dec_exp, constant_float(1.0, spec));
+  if_exprt adjust_for_neg(negative_exp, dec_minus_one, dec_exp);
+  return round_expr_to_zero(adjust_for_neg);
+}
+
+/// Add axioms corresponding to the String.valueOf(F) java function
+/// \param f: function application with one float argument
+/// \return a new string expression
+string_exprt string_constraint_generatort::add_axioms_for_string_of_float(
+  const function_application_exprt &f)
+{
+  return add_axioms_for_string_of_float(
+    args(f, 1)[0], to_refined_string_type(f.type()));
+}
+
+/// Add axioms corresponding to the String.valueOf(D) java function
+/// \param f: function application with one double argument
+/// \return a new string expression
+string_exprt string_constraint_generatort::add_axioms_from_double(
+  const function_application_exprt &f)
+{
+  return add_axioms_for_string_of_float(
+    args(f, 1)[0], to_refined_string_type(f.type()));
+}
+
+/// Add axioms corresponding to the integer part of m, in decimal form with no
+/// leading zeroes, followed by '.' ('\u002E'), followed by one or more decimal
+/// digits representing the fractional part of m.
+///
+/// TODO: this specification is not correct for negative numbers and
```

Comment:
We can't know when adding the axioms whether the number is greater than 10000, the axioms we add ensure that if the input is smaller than 10000 then the output is correct.
But maybe the TODO is misleading here, it should rather be the specification of the function.